File: checks/python/src/section508/checks/error_identification.py
import re

from section508.dom import SPACE, attr, find_id, has_class, select

ERROR_WORDS = re.compile(r"error|missing|invalid|required|must ", re.I)


def check(root):
    if not root:
        return []

    fields = select(root, "input, select, textarea")
    if not fields:
        return ["This section has no field in error."]

    for field in fields:
        if not is_in_error(field):
            continue
        if not error_text(field, root):
            return ["The error is not identified in text."]

    if not any(is_in_error(field) for field in fields):
        return ["The error is not identified in text."]
    return []


def is_in_error(field) -> bool:
    return attr(field, "aria-invalid") == "true" or has_class(field, "invalid")


def error_text(field, root) -> bool:
    ids = [ref for ref in (attr(field, "aria-describedby") or "").split() if ref]
    from_ids = " ".join(find_id(root, ref).get_text() if find_id(root, ref) is not None else "" for ref in ids)
    if SPACE.sub(" ", from_ids).strip():
        return True
    return any(ERROR_WORDS.search(node.get_text() or "") for node in select(root, "p, div, span, strong"))

File: checks/ruby/lib/section508/checks/error_identification.rb
# frozen_string_literal: true

module Section508
  module Checks
    # A field in error is identified in text, not only a CSS class or aria-invalid.
    class ErrorIdentification
      def self.call(root)
        return [] unless root&.respond_to?(:css)

        fields = Dom.elements(root, "input, select, textarea")
        return ["This section has no field in error."] if fields.empty?

        fields.each do |field|
          next unless in_error?(field)
          return ["The error is not identified in text."] unless error_text?(field, root)
        end

        return ["The error is not identified in text."] unless fields.any? { |field| in_error?(field) }

        []
      end

      def self.in_error?(field)
        field["aria-invalid"] == "true" || Dom.has_class?(field, "invalid")
      end

      def self.error_text?(field, root)
        ids = field["aria-describedby"].to_s.split(/\s+/).reject(&:empty?)
        from_ids = ids.map { |id| node = Dom.by_id(root, id); node ? Dom.text(node) : "" }.join(" ")
        return true unless from_ids.gsub(/\s+/, " ").strip.empty?

        Dom.elements(root, "p, div, span, strong").any? do |node|
          /error|missing|invalid|required|must /i.match?(Dom.text(node))
        end
      end
    end
  end
end
